infrastructure/local-provisioners/azurerm_apim.ts

```diff
@@ -0,0 +1,510 @@
+  winston.info(
+    `Replaced configuration parameters in ${JSON.stringify(changes)}`
+  );
+
+  winston.info(
+    "Push local configuration to the master branch of the API management SCM repository"
+  );
+
+  shelljs.exec(`git init .`);
+  shelljs.exec(`git remote add origin ${scmUrlWithCreds}`);
+  shelljs.exec(`git add -A`);
+  shelljs.exec(`git commit -a -m "configuration update"`);
+  shelljs.exec(`git push origin master --force`);
+  shelljs.popd();
+
+  winston.info("Deploy configuration from master branch to API management");
+
+  // Deploy configuration from pushed master branch
+  const deploy = await apiClient.tenantConfiguration.deploy(
+    config.azurerm_resource_group,
+    config.azurerm_apim,
+    {
+      branch: "master",
+      // deletes subscriptions to products that are deleted in this update
+      force: true
+    }
+  );
+  if (deploy.error) {
+    throw new Error(JSON.stringify(deploy));
+  }
+
+  return gitKey;
+};
+
+const getPropsFromFunctions = async (
+  loginCreds: ICreds,
+  config: IResourcesConfiguration
+) => {
+  winston.info("Get Functions application key and backend URL");
+
+  // Get Functions (backend) info
+  // We need these to setup API operations
+  const webSiteClient = new webSiteManagementClient(
+    loginCreds.creds as any,
+    loginCreds.subscriptionId
+  );
+
+  const { masterKey, backendUrl } = await getFunctionsInfo(
+    webSiteClient,
+    config.azurerm_resource_group,
+    config.azurerm_functionapp
+  );
+
+  return {
+    backendUrl: { secret: false, value: backendUrl },
+    code: { secret: true, value: masterKey }
+  };
+};
+
+/**
+ * Creates the API management PaaS
+ */
+const createOrUpdateApiManagementService = async (
+  apiClient: apiManagementClient,
+  config: IResourcesConfiguration
+) => {
+  winston.info(
+    "Create API management resource, this takes a while (about 30 minutes)..."
+  );
+  const apiManagementService = await apiClient.apiManagementService.createOrUpdate(
+    config.azurerm_resource_group,
+    config.azurerm_apim,
+    {
+      location: config.location,
+      notificationSenderEmail: config.apim_email,
+      publisherEmail: config.apim_email,
+      publisherName: config.apim_publisher,
+      sku: { name: config.azurerm_apim_sku, capacity: 1 }
+    }
+  );
+  return apiManagementService;
+};
+
+export const run = async (params: ApimParams) => {
+  const config = readConfig(params.environment).fold(errs => {
+    throw new Error(
+      "Error parsing configuration:\n\n" + reporter(left(errs) as any)
+    );
+  }, t.identity);
+
+  const loginCreds = await login();
+
+  // Needed to get storage connection string
+  const apiClient = new apiManagementClient(
+    loginCreds.creds,
+    loginCreds.subscriptionId
+  );
+
+  const functionProperties = await getPropsFromFunctions(loginCreds, config);
+
+  // Set up backend url and code (master key) named values
+  // to access Functions endpoint in policies
+
+  switch (params.apim_task) {
```

Comment:
there's a little too much going on in this provisioners, I'd like to split it in smaller, simpler, specialized provisioners and rely on terraform to compose them